core/vm/eof.go

```diff
@@ -0,0 +1,340 @@
+			dataRead += 1
+		default:
+			return EOF1Header{}, ErrEOF1UnknownSection
+		}
+		i += 3
+	}
+	// 1 code section is required.
+	if codeRead < 1 {
+		return EOF1Header{}, ErrEOF1CodeSectionMissing
+	}
+	// 1024 max code sections.
+	if len(header.codeSize) > 1024 {
+		return EOF1Header{}, ErrEOF1CodeSectionOverflow
+	}
+	// Must have type section if more than one code section.
+	if len(header.codeSize) > 1 && header.typeSize == 0 {
+		return EOF1Header{}, ErrEOF1TypeSectionMissing
+	}
+	// Declared section sizes must correspond to real size (trailing bytes are not allowed.)
+	if i+int(header.typeSize)+codeSize+int(header.dataSize) != len(code) {
+		return EOF1Header{}, ErrEOF1InvalidTotalSize
+	}
+	return header, nil
+}
+
+// readEOF1Header parses EOF1-formatted code header, assuming that it is already validated.
+func readEOF1Header(code []byte) EOF1Header {
+	var (
+		header       EOF1Header
+		i            = sectionHeaderStart
+		codeSections = uint16(1)
+	)
+	// Try to read type section.
+	if code[i] == byte(kindType) {
+		size, _ := readSectionSize(code, i+1)
+		header.typeSize = size
+		codeSections = size / 2
+		i += 3
+	}
+	i += 1
+	// Read code sections.
+	for j := 0; j < int(codeSections); j++ {
+		size := binary.BigEndian.Uint16(code[i+3*j : i+3*j+2])
+		header.codeSize = append(header.codeSize, size)
+	}
+	i += int(2 * codeSections)
+	// Try to read data section.
+	if code[i] == byte(kindData) {
+		header.dataSize = binary.BigEndian.Uint16(code[i+1 : i+3])
+	}
+	return header
+}
+
+// validateInstructions checks that there're no undefined instructions and code ends with a terminating instruction
+func validateInstructions(code []byte, header *EOF1Header, jumpTable *JumpTable) error {
+	var (
+		i        = 0
+		analysis = codeBitmap(code)
+		opcode   OpCode
+	)
+	for i < len(code) {
+		switch opcode = OpCode(code[i]); {
+		case jumpTable[opcode].undefined:
+			return fmt.Errorf("%v: %v", ErrEOF1UndefinedInstruction, opcode)
+		case opcode >= PUSH1 && opcode <= PUSH32:
+			i += int(opcode) - int(PUSH1) + 2
+			continue // todo make sure this actually continues
+		case opcode == RJUMP || opcode == RJUMPI:
+			var arg int16
+			// Read immediate argument.
+			if err := binary.Read(bytes.NewReader(code[i+1:]), binary.BigEndian, &arg); err != nil {
+				return ErrEOF1InvalidRelativeOffset
+			}
+			// Calculate relative target.
+			pos := i + 3 + int(arg)
+
+			// Check if offset points to out-of-bounds location.
+			if pos < 0 || pos >= len(code) {
+				return ErrEOF1InvalidRelativeOffset
+			}
+			// Check if offset points to non-code segment.
+			// TODO(matt): include CALLF and RJUMPs in analysis.
+			if !analysis.codeSegment(uint64(pos)) {
+				return ErrEOF1InvalidRelativeOffset
+			}
+			i += 3
+		case opcode == CALLF:
+			var arg int16
+			// Read immediate argument.
+			if err := binary.Read(bytes.NewReader(code[i+1:]), binary.BigEndian, &arg); err != nil {
+				return fmt.Errorf("%v: %v", ErrEOF1InvalidCallfSection, err)
+			}
```

Comment:
This looks like a pretty clunky way to read two bytes. This is kind of hot-path, isn't it? Instead of creating readers and using binary.Read which uses io.Readfull and the whole orchestra, I'd just pick out the two bytes and be done.